src/service/viewport/viewport-impl.js

```diff
@@ -396,42 +401,50 @@ export class Viewport {
   }
 
   /**
-   * Returns a promise that resolves with latest rect of the element within the document.
+   * Returns the client rect of the element
+   * TODO(@zhouyx): Consider combine with #getClientRectInViewportAsync()
    * @param {!Element} el
    * @return {!Promise<!../../layout-rect.LayoutRectDef>}
    */
-  getLayoutRectAsync(el) {
-    const scrollLeft = this.getScrollLeft();
-    const scrollTop = this.getScrollTop();
-
-    // Go up the window hierarchy through friendly iframes.
-    const frameElement = getParentWindowFrameElement(el, this.ampdoc.win);
-    if (frameElement) {
-      const bPromise = this.binding_.getLayoutRectAsync(el, 0, 0);
-      const cPromise = this.binding_.getLayoutRectAsync(
-          frameElement, scrollLeft, scrollTop);
-      return Promise.all([bPromise, cPromise]).then(values => {
-        const b = values[0];
-        const c = values[1];
-        return layoutRectLtwh(Math.round(b.left + c.left),
-            Math.round(b.top + c.top),
-            Math.round(b.width),
-            Math.round(b.height));
-      });
-    }
+  getClientRectAsync(el) {
+    const local = this.vsync_.measurePromise(() => {
+      return el./*OK*/getBoundingClientRect();
+    });
+    const global = this.binding_.getGlobalClientRect();
 
-    return this.binding_.getLayoutRectAsync(el, scrollLeft, scrollTop);
+    return Promise.all([local, global]).then(values => {
+      const l = values[0];
+      const g = values[1];
+      if (!g) {
+        return l;
+      }
+      return moveLayoutRect(l, g.left, g.top);
```

Comment:
It's very hard to do anything with this rect w/o knowing the root rect or rather root width and height. Maybe we should consider to return a structure with both? This is a weird thing b/c currently our `Viewport.getRect()` returns an inconsistent value. If we did that, we could also include the intersection box. Between these 3 rects, in theory, the using code can do all necessary inferences. This would be very much similar to InObEvent, wouldn't it?

Reply:
I think this is a good example that we should consider separate `Root` from `Viewport` class. But I honest don't know what I should do here. Now I think returning the clientRect to the global viewport makes sense.